core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala

```diff
@@ -82,12 +86,17 @@ class SchedulerLogicGraph(mesosCallFactory: MesosCalls)
         }
       })
 
-      override def preStart(): Unit = {
-        // Start the stream
-        pull(specEventsInlet)
-        pull(mesosEventsInlet)
+      val startGraph = this.getAsyncCallback[Try[Map[PodId, PodRecord]]] {
+        case Success(initialSnapshot) =>
+          handler.loadInitialPodRecords(initialSnapshot)
+          maybePull()
+        case Failure(ex) =>
+          this.failStage(ex)
       }
 
+      override def preStart(): Unit =
+        initialPodRecords.onComplete(startGraph.invoke)(CallerThreadExecutionContext.context)
```

Comment:
At least in the scope of this PR, I am not trying to handle adhoc-snapshots. Assumption in this PR is that we receive the snapshot only once during bootstrap (or crash recovery) and we intentionally block all inlets until we can load the snapshot and be ready for the processing. I initially had an approach using FanIn3Shape but decided this is simpler and cleaner to get the initial snapshot loading.

Reply:
I think @timcharper has to explain the interface then once more to me. We do not expect adhoc snapshots but it is odd that our internal storage uses plain futures when we prescribe the nested sources to users. Why would we not use streams here as well?

Reply:
In the future, we will not be sending a podSpec snapshot. This is going away.

Reply:
Cool. Could we add a longer comment above?

Reply:
To be clear, removal of podSpec snapshots will not preclude the need for startGraph callback and this preStart code; this will stay.

Reply:
Ah, because this loads, the pod records, right? I still don't quite understand why it is a callback. What do you think of passing `initialPodRecods` to `createLogic`? This would simplify things a bit and the loading of the pod records should not be part of the logic anyways IMHO.

Reply:
Not sure if/how I can pass an extra param to `override def createLogic`. One thought I had was to set an attribute on graph component but it feels like an overkill. Do you have any other thoughts ?

otoh, i don't understand why loading the pod records should not be part of logic - we are creating a graph component which should not start pulling elements from upstream until a certain precondition is satisfied. In our case, the precondition is that the given future completes with an expected (Success) result. Since this future completion is non deterministic (it can even fail), i think it is fair to register the call back in `prestart` method. Ultimately, this callback decides whether (if at all) or not to start pulling elements from upstream.

Reply:
I missed that `createLogic` is predefined. However, `initialPodRecords` could simply be the loaded record set. Why should it be a future?

Reply:
By making pod records a future - we have the option to chose to **either** succeed loading the snapshot _and then_ start the graph **or** mark graph stage as failed to start. If it's not a future, then we have to handle the failed future before creating this graph component (and it brings the question - if we fail to make the persistence call, should we fail to create the blueprint of the graph ?) cc @meln1k

Reply:
Also, note that the `SchedulerLogicHanlder` has a `Map` instead of a `Future[Map]` - https://github.com/mesosphere/usi/pull/80/files#diff-704496199082f7d74e9d62d315f0a7abR66 and it does make sense to me because if we have a "handler", then it can be expected to have a valid snapshot. But in case of `SchedulerLogicGraph` where  we are "building" the graph, we can chose to verify a precondition (in this case - successful completion of the given Future) before we chose to start the graph.

I have made the `initialPodRecords` to be a anon function to delay the pod record loading. If you still think that having a `Future` in graph pre condition is not ideal, I can reformat the code to complete the future in `Scheduler` and fail fast before a graph blue print is generated. wdyt ?

Reply:
I'm just wondering if loading the snapshot in the scheduler factory would be simpler. We can always handle errors there. Ideally the factory method would fail if the snapshot loading would fail. The caller would not even have a failed graph stage. How should the user handle a failed graph stage anyways?